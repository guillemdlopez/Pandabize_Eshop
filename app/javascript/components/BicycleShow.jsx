import React, { useMemo, useState } from "react";
import { Link, useParams } from "react-router-dom";
import { API_BASE_URL } from "../config/variables";
import _ from "lodash";
import { backgroundColor, isColor } from "../config/functions";

const BicycleShow = () => {
  const [bicycle, setBicycle] = useState();
  const { bicycleId } = useParams();

  const getBicycle = async (url) => {
    try {
      const getJSON = await fetch(url);
      const res = await getJSON;
      const data = await res.json();
      if (!data) {
        const message = "Not possible to retrieve data from API";
        throw new Error(message);
      }
      setBicycle(data);
    } catch (error) {
      console.error(error);
    }
  };

  useMemo(() => getBicycle(`${API_BASE_URL}${bicycleId}`), [bicycleId]);

  const wheelSize = bicycle?.customizations.filter(
    (custom) => custom.name === "Wheel size"
  );
  const rimColor = bicycle?.customizations.filter(
    (custom) => custom.name === "Rim color"
  );
  const saddleColor = bicycle?.customizations.filter(
    (custom) => custom.name === "Saddle color"
  );

  return (
    <div className="bicycle-show container mt-5">
      <div className="row align-items-center">
        <div className="col-lg-7 col-md-12  mb-4">
          <img src={bicycle?.photo} className="bicycle-photo" />
        </div>
        <div className="col-lg-5 col-md-12 bicycle-info mb-5">
          <h2 className="bicycle-name">{bicycle?.name}</h2>
          <p className="text-muted bicycle-description">
            {bicycle?.description}
          </p>

          {/* WHEEL SIZE CUSTOMIZATION */}
          {wheelSize?.length === 0 ? (
            ""
          ) : (
            <p className="mb-3 custom-name">
              <strong>Wheel size{wheelSize?.length > 1 ? "s" : ""}</strong>{" "}
            </p>
          )}

          <div className="customization d-flex">
            {wheelSize?.map((size) => {
              return (
                <ul className="inline-list mb-4" key={size.id}>
                  <li className="custom-list-item mr-2">{size.value}</li>
                </ul>
              );
            })}
          </div>

          {/* RIM COLOR CUSTOMIZATION */}
          {rimColor?.length === 0 ? (
            ""
          ) : (
            <p className="mb-0 custom-name">
              <strong>Rim color{rimColor?.length > 1 ? "s" : ""}</strong>
            </p>
          )}

          <div className="customization d-flex mb-3" id="color-customization">
            {rimColor?.map((color) => {
              return (
                <div
                  className="color-value mr-2"
                  style={backgroundColor(color.value)}
                  data-color={color.value}
                  key={color.id}
                ></div>
              );
            })}
          </div>

          {/* SADDLE COLOR CUSTOMIZATION */}
          {saddleColor?.length === 0 ? (
            ""
          ) : (
            <p className="mb-0 custom-name">
              <strong>Saddle color{saddleColor?.length > 1 ? "s" : ""}</strong>
            </p>
          )}
          <div className="customization d-flex mb-3" id="color-customization">
            {saddleColor?.map((color) => {
              return (
                <div
                  className="color-value mr-2"
                  style={backgroundColor(color.value)}
                  data-color={color.value}
                  key={color.id}
                ></div>
              );
            })}
          </div>
          <Link to={`/bicycle/${bicycleId}/order`}>
            <button
              className="btn btn-outline-success btn-block mb-5"
              id="btn-to-form"
            >
              Buy
            </button>
          </Link>
        </div>
      </div>
    </div>
  );
};

export default BicycleShow;
