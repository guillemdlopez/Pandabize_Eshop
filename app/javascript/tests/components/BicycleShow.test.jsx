import React from "react";
import { mount } from "enzyme";
import BicycleShow from "../../components/BicycleShow";
import {
  MemoryRouter,
  BrowserRouter as Route,
  useParams,
} from "react-router-dom";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({ bicycleId: 26 }),
}));

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useMemo: () => `http://localhost:3000/api/v1/bicycles/26`,
}));

describe("Testing the <BicycleShow /> component", () => {
  const { bicycleId } = useParams();

  const wrapper = mount(
    <MemoryRouter initialEntries={[`/bicycle/${bicycleId}`]}>
      <Route path="/bicycle/:bicycleId">
        <BicycleShow />
      </Route>
    </MemoryRouter>
  );
  test("should render the component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should contain a div with the bicycle info", () => {
    expect(wrapper.find(".bicycle-info").exists()).toBe(true);
  });

  test("should contain paragraphs with customizations: Wheel size, Rim color and Saddle Color", () => {
    expect(wrapper.find(".custom-name").at(0).text().trim()).toBe("Wheel size");
    expect(wrapper.find(".custom-name").at(1).text().trim()).toBe("Rim color");
    expect(wrapper.find(".custom-name").at(2).text().trim()).toBe(
      "Saddle color"
    );
  });

  test("should contain a link to the bicycle's order page", () => {
    expect(wrapper.find("a").prop("href")).toBe("/bicycle/26/order");
  });

  test("should contain a button that says 'Buy'", () => {
    expect(wrapper.find("button").text()).toBe("Buy");
  });
});
