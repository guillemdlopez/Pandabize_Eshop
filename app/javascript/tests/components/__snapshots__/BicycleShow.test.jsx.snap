// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the <BicycleShow /> component should render the component correctly 1`] = `
<div
  className="bicycle-show container mt-5 animate__animated animate__fadeIn"
>
  <div
    className="row align-items-center"
  >
    <div
      className="col-lg-7 col-md-12 mb-4"
    >
      <img
        className="bicycle-photo"
      />
    </div>
    <div
      className="col-lg-5 col-md-12 bicycle-info mb-5"
    >
      <h2
        className="bicycle-name"
      />
      <p
        className="text-muted bicycle-description"
      />
      <p
        className="mb-3 custom-name"
      >
        <strong>
          Wheel size
        </strong>
         
      </p>
      <div
        className="customization d-flex"
      />
      <p
        className="mb-0 custom-name"
      >
        <strong>
          Rim color
        </strong>
      </p>
      <div
        className="customization d-flex mb-3"
        id="color-customization"
      />
      <p
        className="mb-0 custom-name"
      >
        <strong>
          Saddle color
        </strong>
      </p>
      <div
        className="customization d-flex mb-3"
        id="color-customization"
      />
      <a
        href="/bicycle/26/order"
        onClick={[Function]}
      >
        <button
          className="btn btn-outline-success btn-block mb-5"
          id="btn-to-form"
        >
          Buy
        </button>
      </a>
    </div>
  </div>
</div>
`;
